src/openms/include/OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h

```diff
@@ -0,0 +1,153 @@
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Lars Nilse $
+// $Authors: Lars Nilse $
+// --------------------------------------------------------------------------
+
+#ifndef OPENMS_FILTERING_DATAREDUCTION_HASHEDSPECTRUM_H
+#define OPENMS_FILTERING_DATAREDUCTION_HASHEDSPECTRUM_H
+
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/DATASTRUCTURES/DRange.h>
+
+#include <vector>
+
+namespace OpenMS
+{
+/**
+ * @brief Auxiliary data structure for the quick access of peaks in MSSpectrum<Peak1D>
+ *
+ * Finding a peak with a particular m/z in MSSpectrum<Peak1D> can be slow. This data structure
+ * provides a hash index (bins_) for quick access. Given an m/z of interest, one can
+ * quickly jump in the relevant neighbouthood i.e. bin(s) where such a peak might be located.
+ * The getPeak(double mz) method returns a pointer to the peak in MSSpectrum<Peak1D>.
+ *
+ * @see MSSpectrum<Peak1D>
+ */
+class OPENMS_DLLAPI HashedSpectrum
+{
+  public:
+  
+    /**
+     * @brief m/z interval containing pointers to the first and last data points
+     * If there are no data points in this interval, both pointers are null.
+     * Boundaries of the interval implicitly given by mz_min, mz_bin, mz_unit_ppm_
+     * and the hash table index.
+     */
+    struct MzInterval
+    {
+		MSSpectrum<Peak1D>::pointer first;
+		MSSpectrum<Peak1D>::pointer last;
+	};
+    
+    /**
+     * @brief constructor taking an MSSpectrum
+     * 
+     * @param raw_spectrum    raw MS1 spectrum to be indexed
+     * @param mz_bin    size of the m/z bins aka intervals
+     * @param mz_tolerance    maximal difference between m/z of interest and m/z of a valid peak
+     * @param mz_unit_ppm    unit for mz_bin and mz_tolerance
+     */
+    HashedSpectrum(MSSpectrum<Peak1D>& raw_spectrum, double mz_bin, double mz_tolerance, bool mz_unit_ppm);
+
+    /**
+     * @brief destructor
+     */
+    ~HashedSpectrum();
+
+    /**
+     * @brief returns the m/z bin size
+     */
+    double getMzBin() const;
```

Comment:
Bin size aka interval size is the m/z range associated with each each hash index. Used in the constructor.

Tolerance is the maximum difference between the mz of interest and the mz of a peak that will be returned. Typically smaller than bin size. Used in `getPeak()`.

https://github.com/OpenMS/OpenMS/pull/2377/files#diff-91a1b4a0d6dfa5b0bbbef40227ccea7aR75